File: app/src/main/java/rogue_opcode/Spritey.java
// Spritey.java
// Sprite class with sheet/frame animation and collision
//
// Copyright © 2009-2017 Brigham Toskin, Leea Harlan
// This software is part of the Rogue-Opcode game framework. It is distributable
// under the terms of a modified MIT License. You should have received a copy of
// the license in the file LICENSE.md. If not, see:
// <http://code.google.com/p/rogue-op/wiki/LICENSE>
//
// Formatting:
//	80 cols ; tabwidth 4
////////////////////////////////////////////////////////////////////////////////


package rogue_opcode;

import android.graphics.Canvas;
import android.graphics.RectF;

import rogue_opcode.geometrics.Rectangle;
import rogue_opcode.geometrics.Rectanglef;
import rogue_opcode.geometrics.XYf;


/**
 * Extends {@code ScreenElement} with sprite strip animation, and collision.
 *
 * @see ScreenElement
 * @author Brigham Toskin
 */
public class Spritey extends ScreenElement
{
	private static final long serialVersionUID = -5463549117882033160L;

	protected Rectangle mFrame;
	protected int mFrameWidth;
	protected int mFrameHeight;
	protected int mFrameCount;
	public int msecsPerFrame = 33;	// ~30 FPS
	public int timeSinceLastFrame = 0;

	public Rectanglef boundingBox;
	public int bbOffset = 0;

	// c'tor ///////////////////////////////////////////////////////////////////

	/**
	 * Constructs a Spritey to load the specified resource.
	 * 
	 * @param pResourceID resource to load.
	 * @param pFrameCount number of frames in the sprite strip.
	 */
	public Spritey(int pResourceID, int pFrameCount)
	{
		super(pResourceID);
		init(pFrameCount);
	}

	/**
	 * Constructs a Spritey to load the specified resource at specified logical
	 * screen coordinates.
	 * 
	 * @param pResourceID resource to load.
	 * @param pFrameCount number of frames in the sprite strip.
	 * @param pX horizontal position of new Spritey.
	 * @param pY vertical position of new Spritey.
	 */
	public Spritey(int pResourceID, int pFrameCount, int pX, int pY)
	{
		super(pResourceID, pX, pY);
		init(pFrameCount);
	}

	protected void init(int pFrameCount)
	{
		Hibernate();
		mFrameCount = pFrameCount;
		mFrameWidth = mGR.PhysicalWidth() / pFrameCount;
		mFrameHeight = mGR.PhysicalHeight();
		mFrame = new Rectangle(0, 0, mFrameWidth, mFrameHeight);
	}

	// game loop callbacks /////////////////////////////////////////////////////

	/**
	 * Handles frame animation and velocity-based displacement updates.
	 *
	 * @see rogue_opcode.ScreenElement#Update()
	 */
	@Override
	public void Update()
	{
		Rectangle tFrame = mFrame;
		timeSinceLastFrame += GameProc.UPDATE_PERIOD;
		while(timeSinceLastFrame > msecsPerFrame)
		{
			timeSinceLastFrame -= msecsPerFrame;
			tFrame.X(mFrameWidth);
			if(tFrame.X() >= mGR.PhysicalWidth())
				// right side is off the end; reset
				resetFrame();
		}
		super.Update();
	}

	@Override
	public void Draw()
	{
		Canvas tCanvas = AnimatedView.sCurrentCanvas;
		if(tCanvas != null)
		{
			XYf tPos = mPos;
			RectF tDest = new RectF(tPos.x, tPos.y, tPos.x + mFrameWidth,
					tPos.y + mGR.PhysicalHeight());
			tCanvas.drawBitmap(mGR.mImage, mFrame.toRect(), tDest, null);
		}
	}

	protected void resetFrame()
	{
		mFrame.X(0);
	}

	/**
	 * Checks for collision between two {@code Spritey}s.
	 *
	 * @param pSpritey {@code Spritey} to check against
	 * @return {@code true} on collision, {@code false} otherwise
	 */
	public boolean Collide(Spritey pSpritey)
	{
		return Collide(pSpritey.boundingBox);
	}

	/**
	 * Checks for collision with an arbitrary {@link Rectangle}.
	 *
	 * @param pBBox bounding box of {@code Spritey} to check against
	 * @return {@code true} on collision, {@code false} otherwise
	 */
	public boolean Collide(Rectanglef pBBox)
	{
		return boundingBox.toRectF().intersect(pBBox.toRectF());
	}

	// Spritey properties //////////////////////////////////////////////////////

	// Translates between frames per second and delay between frames.
	public int FrameRate()
	{
		return 1000 / msecsPerFrame;
	}

	public void FrameRate(int value)
	{
		msecsPerFrame = 1000 / value;
	}

	// Gets or sets the collision bounding box offset
	public int Offset()
	{
		return bbOffset;
	}

	public void Offset(int value)
	{
		bbOffset = value;
		boundingBox.X((int)mPos.x + bbOffset);
		boundingBox.Y((int)mPos.y + bbOffset);
		boundingBox.W(mFrameWidth - bbOffset * 2); // shrink in from both sides
		boundingBox.H(mFrameHeight - bbOffset * 2);
	}


	/**
	 * Gets the current bounding box.
	 *
	 * @return the current bounding box.
	 */
	public Rectanglef BoundingBox()
	{
		return boundingBox;
	}

	public void BoundingBox(Rectanglef box)
	{
		boundingBox = box;
	}
}

File: lib/src/main/java/rogue_opcode/Spritey.kt
// Spritey.kt
// Sprite class with sheet/frame animation and collision
//
// Copyright © 2009-2017 Brigham Toskin, Leea Harlan
// This software is part of the Rogue-Opcode game framework. It is distributable
// under the terms of a modified MIT License. You should have received a copy of
// the license in the file LICENSE.md. If not, see:
// <https://github.com/IonoclastBrigham/RogueOp/blob/master/LICENSE.md>
//
// Formatting:
//	80 cols ; tabwidth 4
////////////////////////////////////////////////////////////////////////////////


package rogue_opcode

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import rogue_opcode.geometrics.Rectangle
import rogue_opcode.geometrics.Rectanglef


/**
 * Extends `ScreenElement` with sprite strip animation, and collision.

 * @see ScreenElement

 * @author Brigham Toskin
 */
open class Spritey : ScreenElement {
	companion object {
		private val serialVersionUID = -5463549117882033160L
	}


	protected var mFrame = Rectangle()
	protected var mFrameWidth: Int = 0
	protected var mFrameHeight: Int = 0
	protected var mFrameCount: Int = 0
	protected var msecsPerFrame = 33	// ~30 FPS
	protected var timeSinceLastFrame = 0

	var BoundingBox = Rectanglef()

	// Translates between frames per second and delay between frames.
	var FrameRate
		get() = 1000 / msecsPerFrame
		set(value) { msecsPerFrame = 1000 / value }

	// Gets or sets the collision bounding box offset
	var Offset = 0
		set(value) {
			field = value
			BoundingBox.X((mPos.x.toInt() + field).toFloat())
			BoundingBox.Y((mPos.y.toInt() + field).toFloat())
			BoundingBox.W((mFrameWidth - field * 2).toFloat()) // shrink in from both sides
			BoundingBox.H((mFrameHeight - field * 2).toFloat())
		}

	// c'tor ///////////////////////////////////////////////////////////////////

	/**
	 * Constructs a Spritey to load the specified resource.
	 *
	 * @param pResourceID framestrip resource to load.
	 * @param pFrameCount number of frames in the sprite strip.
	 */
	constructor(pResourceID: Int, pFrameCount: Int) : super(pResourceID) {
		init(pFrameCount)
	}

	/**
	 * Constructs a Spritey to load the specified resource at specified logical
	 * screen coordinates.
	 *
	 * @param pResourceID framestrip resource to load.
	 * @param pFrameCount number of frames in the sprite strip.
	 * @param pX horizontal position of new Spritey.
	 * @param pY vertical position of new Spritey.
	 */
	constructor(pResourceID: Int, pFrameCount: Int, pX: Int, pY: Int) : super(pResourceID, pX, pY) {
		init(pFrameCount)
	}

	protected fun init(pFrameCount: Int) {
		Hibernate()
		mFrameCount = pFrameCount
		mFrameWidth = currentGR!!.PhysicalWidth() / pFrameCount
		mFrameHeight = currentGR!!.PhysicalHeight()
		mFrame = Rectangle(0, 0, mFrameWidth, mFrameHeight)
	}

	// game loop callbacks /////////////////////////////////////////////////////

	/**
	 * Handles frame animation and velocity-based displacement updates.
	 *
	 * @see rogue_opcode.ScreenElement.Update
	 */
	override fun Update() {
		val tFrame = mFrame
		timeSinceLastFrame += GameProc.UPDATE_PERIOD.toInt()
		while(timeSinceLastFrame > msecsPerFrame) {
			timeSinceLastFrame -= msecsPerFrame
			tFrame.X(mFrameWidth)
			if(tFrame.X() >= currentGR!!.PhysicalWidth())
			// right side is off the end; reset
				resetFrame()
		}
		super.Update()
	}

	override fun Draw(pCanvas: Canvas, pPreScaler: Float, pDefaultPaint: Paint) {
		val (tX, tY) = mPos
		val tDest = RectF(tX, tY, tX + mFrameWidth, tY + currentGR!!.PhysicalHeight())
		pCanvas.drawBitmap(currentGR!!.mImage!!, mFrame.toRect(), tDest, pDefaultPaint)
	}

	protected fun resetFrame() {
		mFrame.X(0)
	}

	/**
	 * Checks for collision between two `Spritey`s.
	 *
	 * @param pSpritey `Spritey` to check against
	 * @return `true` on collision, `false` otherwise
	 */
	fun Collide(pSpritey: Spritey): Boolean {
		return Collide(pSpritey.BoundingBox)
	}

	/**
	 * Checks for collision with an arbitrary [Rectangle].
	 *
	 * @param pBBox bounding box of `Spritey` to check against
	 * @return `true` on collision, `false` otherwise
	 */
	fun Collide(pBBox: Rectanglef): Boolean {
		return BoundingBox.toRectF().intersect(pBBox.toRectF())
	}
}
